File: electron/database/templateRepository.ts
import type Database from "better-sqlite3";
import type { Template } from "../types";
import { prepareSoftDeleteStatements } from "./repositoryHelpers";

interface TemplateRow {
  id: string;
  name: string;
  content: string;
  is_deleted: number;
  deleted_at: string | null;
  version: number;
  created_at: string;
  updated_at: string;
}

function rowToNode(row: TemplateRow): Template {
  return {
    id: row.id,
    name: row.name,
    content: row.content,
    isDeleted: row.is_deleted === 1,
    deletedAt: row.deleted_at ?? undefined,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function createTemplateRepository(db: Database.Database) {
  const stmts = {
    fetchAll: db.prepare(
      `SELECT * FROM templates WHERE is_deleted = 0 ORDER BY created_at ASC`,
    ),
    fetchById: db.prepare(`SELECT * FROM templates WHERE id = ?`),
    insert: db.prepare(`
      INSERT INTO templates (id, name, content, is_deleted, created_at, updated_at)
      VALUES (@id, @name, '', 0, datetime('now'), datetime('now'))
    `),
    update: db.prepare(`
      UPDATE templates
      SET name = COALESCE(@name, name),
          content = COALESCE(@content, content),
          version = version + 1,
          updated_at = datetime('now')
      WHERE id = @id
    `),
    ...prepareSoftDeleteStatements(db, "templates"),
  };

  return {
    fetchAll(): Template[] {
      return (stmts.fetchAll.all() as TemplateRow[]).map(rowToNode);
    },

    fetchById(id: string): Template | undefined {
      const row = stmts.fetchById.get(id) as TemplateRow | undefined;
      return row ? rowToNode(row) : undefined;
    },

    create(id: string, name: string): Template {
      stmts.insert.run({ id, name });
      const row = stmts.fetchById.get(id) as TemplateRow;
      return rowToNode(row);
    },

    update(id: string, updates: { name?: string; content?: string }): Template {
      stmts.update.run({
        id,
        name: updates.name ?? null,
        content: updates.content ?? null,
      });
      const row = stmts.fetchById.get(id) as TemplateRow;
      return rowToNode(row);
    },

    softDelete(id: string): void {
      stmts.softDelete.run(id);
    },

    restore(id: string): void {
      stmts.restore.run(id);
    },

    permanentDelete(id: string): void {
      stmts.permanentDelete.run(id);
    },
  };
}

export type TemplateRepository = ReturnType<typeof createTemplateRepository>;

File: src-tauri/src/db/template_repository.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::helpers;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Template {
    pub id: String,
    pub name: String,
    pub content: String,
    pub is_deleted: bool,
    pub deleted_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

fn row_to_template(row: &rusqlite::Row) -> rusqlite::Result<Template> {
    Ok(Template {
        id: row.get("id")?,
        name: row.get("name")?,
        content: row.get("content")?,
        is_deleted: row.get::<_, i64>("is_deleted").map(|v| v != 0)?,
        deleted_at: row.get("deleted_at")?,
        created_at: row.get("created_at")?,
        updated_at: row.get("updated_at")?,
    })
}

pub fn fetch_all(conn: &Connection) -> rusqlite::Result<Vec<Template>> {
    let mut stmt = conn.prepare(
        "SELECT * FROM templates WHERE is_deleted = 0 ORDER BY created_at",
    )?;
    let rows = stmt.query_map([], |row| row_to_template(row))?;
    rows.collect()
}

pub fn fetch_by_id(conn: &Connection, id: &str) -> rusqlite::Result<Option<Template>> {
    let mut stmt = conn.prepare("SELECT * FROM templates WHERE id = ?1")?;
    let result = stmt.query_row([id], |row| row_to_template(row));
    match result {
        Ok(template) => Ok(Some(template)),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e),
    }
}

pub fn create(conn: &Connection, id: &str, name: &str) -> rusqlite::Result<Template> {
    let now = helpers::now();
    conn.execute(
        "INSERT INTO templates (id, name, content, is_deleted, version, created_at, updated_at) \
         VALUES (?1, ?2, '', 0, 1, ?3, ?4)",
        params![id, name, &now, &now],
    )?;

    let mut stmt = conn.prepare("SELECT * FROM templates WHERE id = ?1")?;
    stmt.query_row([id], |row| row_to_template(row))
}

pub fn update(conn: &Connection, id: &str, updates: &Value) -> rusqlite::Result<Template> {
    let mut sets = Vec::new();
    let mut values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

    if let Some(v) = updates.get("name") {
        sets.push("name = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }
    if let Some(v) = updates.get("content") {
        sets.push("content = ?");
        values.push(Box::new(v.as_str().map(|s| s.to_string())));
    }

    if sets.is_empty() {
        let mut stmt = conn.prepare("SELECT * FROM templates WHERE id = ?1")?;
        return stmt.query_row([id], |row| row_to_template(row));
    }

    sets.push("version = version + 1");
    sets.push("updated_at = datetime('now')");
    values.push(Box::new(id.to_string()));

    let sql = format!(
        "UPDATE templates SET {} WHERE id = ?",
        sets.join(", ")
    );
    let params: Vec<&dyn rusqlite::types::ToSql> = values.iter().map(|v| v.as_ref()).collect();
    conn.execute(&sql, params.as_slice())?;

    let mut stmt = conn.prepare("SELECT * FROM templates WHERE id = ?1")?;
    stmt.query_row([id], |row| row_to_template(row))
}

pub fn soft_delete(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    helpers::soft_delete(conn, "templates", id)
}

pub fn permanent_delete(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    helpers::permanent_delete(conn, "templates", id)
}
